src/jit/codegenxarch.cpp

```diff
@@ -6841,37 +6908,79 @@ void        CodeGen::genCompareLong(GenTreePtr  treeNode)
     getEmitter()->emitInsBinary(ins, cmpAttr, hiOp1, hiOp2);
 
     // Generate the first jump for the high compare
-    emitJumpKind jumpKind[2];
-    bool branchToTrueLabel[2];
-    genJumpKindsForTreeLongHi(tree, jumpKind, branchToTrueLabel);
-
-    BasicBlock* label = genCreateTempLabel();
-    inst_JMP(jumpKind[0], label);
+    CompareKind compareKind = ((tree->gtFlags & GTF_UNSIGNED) != 0) ? CK_UNSIGNED : CK_SIGNED;
+    
+    // TODO-X86-CQ: When targetReg == REG_NA, if the next block is a BBJ_ALWAYS, we can set
+    //              labelFinal = compiler->compCurBB->bbNext->bbJumpDest.
+    BasicBlock* labelHi = genCreateTempLabel();
+    BasicBlock* labelFinal = genCreateTempLabel();
 
-    // Generate the second jump for LE, LT, GT, and GE. We only do the lower compare if
-    // the hi parts are equal
-    if (jumpKind[1] != EJ_NONE)
+    if (compareKind == CK_SIGNED && (tree->gtOper != GT_NE && tree->gtOper != GT_EQ))
     {
-        assert(branchToTrueLabel[1]);
-        inst_JMP(jumpKind[1], label);
-    }
+        // If we are doing a signed comparison, we need to do a signed set if the high compare is true,
+        // but an unsigned set if we fall through to the low compare. If we have a GT_NE or GT_EQ, we do not
+        // need to worry about the sign of the comparison, so we can use the simplified case.
 
-    // Now create compare for low parts
-    ins = INS_cmp;
-    cmpType = TYP_INT;
-    cmpAttr = emitTypeSize(cmpType);
+        // We only have to check for equality for the hi comparison. If they are not equal, then the set will
+        // do the right thing. If they are equal, we have to check the lo halves.
+        inst_JMP(EJ_jne, labelHi);
 
-    // Emit the comparison
-    getEmitter()->emitInsBinary(ins, cmpAttr, loOp1, loOp2);
+        // Now create compare for low parts
+        ins = INS_cmp;
+        cmpType = TYP_INT;
+        cmpAttr = emitTypeSize(cmpType);
 
```

Comment:
Same for the "else" case, below.

---
In reply to: [71197887](https://github.com/dotnet/coreclr/pull/6219#discussion_r71197887) [](ancestors = 71197887)